feature/ND/230/Node-Event-Scheduler

ShaneTWilliams opened the pull request:
Welcome to ghetto RTOS, folks

`EventScheduler` is responsible for calling callback functions at given intervals, essentially creating a system of (many) non-blocking delays. It's very heavily based on the Arduino timer we used last year.

The `EventScheduler` class owns an array of 50 `Event` objects, each of which has a delay, a count of how many times it needs to be called, a type, a state, the last time it was called, and a callback function. Through the `EventScheduler` class, you can set one of these `Event`s to run a callback every X seconds. 50 is an arbitrary number which can be changed in the future.

`EventScheduler` has 3 methods that matter:
- `EventScheduler::callFunctionAfter` calls the callback once after X seconds
- `EventScheduler::callFunctionEvery` (version 1) calls the callback every X seconds
- `EventScheduler::callFunctionEvery` (version 2) calls the callback every X seconds, but only up until the specified number of calls has been reached

Each of these returns an `eventId_t ` ID which is used to identify that event, and can be handy when using the class's `startEvent` and `stopEvent` methods. Also, whenever you call one of those 3 methods, the class automatically assigns you an unused event object, and therefore an unused event ID.

Every loop in the main, you **_NEED_** to call the `EventScheduler::updateEvents` function. This is what checks every event to see if its delay has passed. If so, the event's callback is called. If the event has been run its maximum number of times, its `Event::cleanup` function is called, and it resets all its properties and becomes ready to be assigned again.

There's also unit tests under `OnPod/Node/Test/EventScheduler`. Pls review them too.

P.S. - the `Event` class definitely could have been been made a struct, but the current solution is super clean and leaves room for the future if needed.

ShaneTWilliams committed:
Add untested EventScheduler class

ShaneTWilliams committed:
Add untested Event class

ShaneTWilliams committed:
Fix small i2c interface error for compiling purposes

ShaneTWilliams committed:
Remove local variable `event` and operate on member array of events

ShaneTWilliams committed:
Add setter for m_num_calls_left

ShaneTWilliams committed:
Change EventScheduler::callFunctionAfter behaviour to be callFunctionEvery repeated one time

ShaneTWilliams committed:
Remove example test

ShaneTWilliams committed:
Add more descriptive comments to EventScheduler class

ShaneTWilliams committed:
Add more descriptive comments to Event class

ShaneTWilliams committed:
Update platformio.ini to include event scheduler

ShaneTWilliams committed:
Change order of callFunctionEvery args

ShaneTWilliams committed:
Add comments and whitespace fixes

ShaneTWilliams committed:
Add testing main for EventScheduler class

ShaneTWilliams committed:
Remove Testing as default env before PR

colton-smith commented on OnPod/Node/Common/EventScheduler/include/event.h:
```diff
@@ -0,0 +1,54 @@
+        void callCallback(void);
+        // Reset event attributes to prepare for reassignment
+        void cleanup(void);
+};
```
This is really clean - but I wonder if making the `EventScheduler` class a friend of the `Event` class would be a cleaner way of doing it. Looks as though the `Event` class is only used by the `EventScheduler` class, not anywhere else. Using a friend class would replace all the getters and setters by providing the `EventScheduler` class access to the `Event` classes private members.  Might be less robust that way though. Let me know what you think.

colton-smith replied:
On second thought, that would detract from encapsulation and also require too much unnecessary work.

colton-smith commented on OnPod/Node/Test/EventScheduler/main.cpp:
```diff
@@ -0,0 +1,253 @@
+}
+
+// stopEvent
+void stopEvent_eventStopped_eventStaysStopped(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionAfter(1000, testCallback, 0);
+    event_scheduler.stopEvent(event);
+    event_scheduler.stopEvent(event);
+    event_scheduler.updateEvents(1001);
+    TEST_ASSERT_EQUAL_UINT16(0, test_callback_run_count);
+    resetTestCallback();
+}
```
`stopEvent` occurs twice consecutively here on lines 45 and 46, intentional?

colton-smith commented on OnPod/Node/Test/EventScheduler/main.cpp:
```diff
@@ -0,0 +1,253 @@
+    event_scheduler.updateEvents(1001);
+    TEST_ASSERT_EQUAL_UINT16(1, test_callback_run_count);
+}
+
```
Do you ever test the case where there are no available event_ids? The case where the `findFreeEvent` methods returns -1?

ShaneTWilliams replied:
Yeah I'll add this. Good catch

Newfloridadad replied:
If the -1 is returned I would send a can message/ set a flag in a node health message if we have one to indicate the event buffer is full. Or alternatively, use a vector if it works on the hardware for flexibility overall but I am fine with the array implementation with a max capacity check.

colton-smith commented on OnPod/Node/Test/EventScheduler/main.cpp:
```diff
@@ -0,0 +1,253 @@
+    resetTestCallback();
+}
+
+void callFunctionEveryFinite_delayPassedThreeTimes_functionCalledTwice(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionEvery(1000, testCallback, 0, 2);
+    event_scheduler.updateEvents(1001);
+    event_scheduler.updateEvents(2002);
+    event_scheduler.updateEvents(3003);
+    TEST_ASSERT_EQUAL_UINT16(2, test_callback_run_count);
+    resetTestCallback();
+}
```
Maybe repeat this test except create a new event and ensure the event_id is 0 and not 1? To make sure the old complete event is properly cleared for use. Just to check off all the boxes?

ShaneTWilliams replied:
I'm kinda confused about what you mean here

colton-smith replied:
I just mean maybe test that finite events get cleaned up after they "expire", could be too explicit of a test though.

Newfloridadad replied:
store the id returned by event 1000, let it expire. Create event 2000 and the id should be the same as 1000, both being 0. this ensures 1000 was cleaned up and replaced properly.

Additionally, I would fill with 51 events to ensure the -1 test recommended in coltons other comment works, have 1 expire then try to add a 52 which should succed. This to check an edge case around the max event buffer size.

colton-smith reviewed (commented):
Loved it.
Style-wise it's very clean, no complaints here.

Although I don't have much experience with event-driven code, I'm able to understand this which is a good sign in terms of readability. Code-wise the only thing I saw is the possibility of making the `EventScheduler` class a friend of the `Event` class, as I mentioned in the comment. 

Testing is thorough - love to see it. I added a few ideas for test cases, the more boxes we check the better. 

Great job fearless leader

andrewnash commented on OnPod/Node/Common/EventScheduler/src/event.cpp:
```diff
@@ -0,0 +1,78 @@
+void Event::callCallback(void)
+{
+    m_callback();
+}
```
Basically, everything in this file is just one line returns. Would be much easier and cleaner to inline this in the header file

ShaneTWilliams replied:
I thought that was bad practice?

ShaneTWilliams replied:
If it's not a big deal then I'd rather change it to what you suggested cause it definitely is cleaner

andrewnash replied:
it's not bad practice when the implementation is very straight forward (one-line command) like this

andrewnash commented on OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp:
```diff
@@ -0,0 +1,85 @@
+
+eventId_t EventScheduler::findFreeEvent(void)
+{
+    for (eventId_t id = 0; id < m_MAX_NUM_EVENTS; id++)
```
It would literally be faster to check cells randomly than to linear search every time in practice. Any time you have to linear search, you should be rethinking your data structure to make sure it's the best choice, consider using a hash table or using another data structure to keep track of free event ids.

ShaneTWilliams replied:
I figured that iterating through 50 items then making one comparison isn't really that time consuming so I didn't bother but you're definitely right

ShaneTWilliams replied:
> Maybe we are missing something but unsure how this is non-blocking, pretty sure would need multiple threads to accomplish this? Should these events have some priority based on the time elapsed? Maybe some form of priority queue is better suited?

This code isn't blocking, it only checks for events due when you call the eventUpdate() function. I don't think that priorities are really worth the hassle

andrewnash replied:
Non-blocking refers to code that doesn't block the execution of other code. In this case, calling `callCallback` one "event" blocks calling `callCallback` of all other events until complete.

Also as a side note, after writing that sentence, I am triggered your method naming  `callCallback`.

ShaneTWilliams replied:
Yes obviously these function calls are going to block others, the plan was never to have threading on the nodes unless we really needed it. I moreso meant that this class replaces blocking `delay()` calls in our main.

Also yeah lmao at that name but it is describing exactly what the method does

andrewnash replied:
the word call is redundant, by calling a method it is implied that you are... calling it. I understand that you want to represent that the method is a wrapper but I don't think that info is necessary in the name

ShaneTWilliams replied:
So I name the method `callback()`? That's the name of the attribute that holds the pointer to the callback function. What I have should be fine

andrewnash replied:
`callCallback` calls `m_callback`, don't see anywhere where the name `callback` is taken?

ShaneTWilliams replied:
I'll change it

andrewnash replied:
Regarding the original comment about linear searching, are you able to use `std::unordered_map` or something similar on the node? I assume you can't use vectors, otherwise, you would have already.

ShaneTWilliams replied:
I'm 99% sure I could but dynamic memory is sorta taboo with embedded applications and I've been trying to avoid it

ShaneTWilliams replied:
@stuckatmarine any insight?

Newfloridadad replied:
They can cause a crash if you run out of memory, typically MB or KB which I believe the STM is in. If you cap a hashmap/vectors size youll be fine and shave some iterations. Sometimes microcontrolls are finnicy on if they support them at all though. a secondary set/ stack of open event indexes would save you on lookup, not run over size and provide a running count of free event spaces if you did want to stick with an array as the main structure.
All you would have to do is remove values when adding events and add it back when in cleanup(), but realistically youve built a priority queue based on time.

ShaneTWilliams replied:
Are you suggesting making that secondary array dynamic?

Newfloridadad replied:
I wouldnt use a second array but a stack or a set which are inhetintly dynamic, but fill them with all 0 to 'num of event ids' size on creation. They dont grow any bigger then that size so running out of memory wont be an issue and give u O(1) functionality.

andrewnash replied:
On the other hand, opting to pivot to a dynamic data struct removes the need for the second array. I'm a little spooked about undefined behaviour if the event scheduler runs out of room in its array, possible edge case if another class was accidentally adding more events than it needed.

ShaneTWilliams replied:
Ok that makes more sense. Thanks

ShaneTWilliams replied:
@andrewnash I think that returning an error and dealing with that accordingly is safer than using dynamic memory. STL stuff is rarely used in embedded applications

andrewnash replied:
![image](https://user-images.githubusercontent.com/33398082/68727817-0a10c980-05a0-11ea-9597-0fc8716dce6f.png)

To my knowledge, there is much debate between people far smarter than both of us as to whether STL is appropriate in embedded applications. Also STL != dynamic memory. Regardless of whether you use STL vectors or plain C arrays, you can still use sort(), binary_search(), random_shuffle(), the functions for building and managing heaps, etc these will almost certainly be faster and less buggy than versions you build yourself.... like a linear search for loop

ShaneTWilliams replied:
Fair points 100% but given the fact that I have 4 exams this week I won't be able to make changes to the code anyways until this weekend haha. Let's shelf this until then

andrewnash reviewed (changes requested):
These comments are an amalgamation of me and Mark briefly looking at this together. Maybe we are missing something but unsure how this is non-blocking, pretty sure would need multiple threads to accomplish this? Should these events have some priority based on the time elapsed? Maybe some form of priority queue is better suited?

Burke-Daniel reviewed (commented):
Looks great!

Style is clean and consistent, and I am a fan of the test function names, it makes it very explicit and easy to read.

Like Colton I don't have much experience with event driven code, however I understand what this code accomplishes, and anything I noticed has already been addessed by other's comments.

Great job!

Newfloridadad commented on OnPod/Node/Common/EventScheduler/include/event.h:
```diff
@@ -0,0 +1,54 @@
+        void cleanup(void);
+};
+
+#endif
```
very minor, gits picking up your lack of newline at eof

ShaneTWilliams committed:
Add newlines to end of files

ShaneTWilliams committed:
Move implementation of `Event` methods to header

ShaneTWilliams committed:
Rename `Event` method `callCallback` to `callback` :(

ShaneTWilliams committed:
Change `Event` class to a struct and change m_events from an array to a STL map

ShaneTWilliams committed:
Change eventId_t from int8_t to int16_t in case if more than 127 events are needed

ShaneTWilliams committed:
Add getter for m_MAX_NUM_EVENTS

ShaneTWilliams committed:
Add unit tests suggested in PR

ShaneTWilliams committed:
Add new comments to EventScheduler class

ShaneTWilliams committed:
Add better comments to `findFreeEventId`

ShaneTWilliams committed:
Move EventScheduler from Node/Common to OnPod/Common

ShaneTWilliams commented:
Changed the main data structure to a map. It's just as fast (as a vector) when iterating and faster in finding available eventId, inserting and erasing.

Changed the `Event` class to a struct because it was only holding data anyways.

Added the tests suggested in the PR.

Moved the whole rig to OnPod/Common from Node/Common.
